datafusion/core/src/physical_plan/insert.rs

```diff
@@ -0,0 +1,203 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Execution plan for writing data to [`DataSink`]s
+
+use super::expressions::PhysicalSortExpr;
+use super::{
+    DisplayFormatType, ExecutionPlan, Partitioning, SendableRecordBatchStream, Statistics,
+};
+use crate::error::Result;
+use arrow::datatypes::SchemaRef;
+use arrow::record_batch::RecordBatch;
+use arrow_array::{ArrayRef, UInt64Array};
+use arrow_schema::{DataType, Field, Schema};
+use async_trait::async_trait;
+use core::fmt;
+use futures::StreamExt;
+use std::any::Any;
+use std::sync::Arc;
+
+use crate::execution::context::TaskContext;
+use crate::physical_plan::stream::RecordBatchStreamAdapter;
+use crate::physical_plan::Distribution;
+use datafusion_common::DataFusionError;
+
+/// `DataSink` implements writing streams of [`RecordBatch`]es to
+/// destinations.
+#[async_trait]
+pub trait DataSink: std::fmt::Debug + Send + Sync {
+    // TODO add desired input ordering
+    // How does this sink want its input ordered?
+
+    /// Writes the data to the sink, returns the number of values written
+    ///
+    /// This method will be called exactly once during each DML
+    /// statement. Thus prior to return, the sink should do any commit
+    /// or rollback required.
+    async fn write_all(&self, data: SendableRecordBatchStream) -> Result<u64>;
+}
+
+/// Execution plan for writing record batches to a [`DataSink`]
+///
+/// Returns a single row with the number of values written
+pub struct InsertExec {
+    /// Input plan that produces the record batches to be written.
+    input: Arc<dyn ExecutionPlan>,
+    /// Sink to whic to write
+    sink: Arc<dyn DataSink>,
+    /// Schema describing the structure of the data.
+    schema: SchemaRef,
+}
+
+impl fmt::Debug for InsertExec {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "InsertExec schema: {:?}", self.schema)
+    }
+}
+
+impl InsertExec {
+    /// Create a plan to write to `sink`
+    pub fn new(input: Arc<dyn ExecutionPlan>, sink: Arc<dyn DataSink>) -> Self {
+        Self {
+            input,
+            sink,
+            schema: make_count_schema(),
+        }
+    }
+}
+
+impl ExecutionPlan for InsertExec {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    /// Get the schema for this execution plan
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        None
+    }
+
+    fn required_input_distribution(&self) -> Vec<Distribution> {
+        vec![Distribution::SinglePartition]
+    }
+
+    fn maintains_input_order(&self) -> Vec<bool> {
+        vec![false]
```

Comment:
🤔  

I am thinking that `InsertExec` is going to be the top of the plan (I don't expect it feeds its output to another `ExecutionPlan`, and produces metadata about the write (the `count`) so I don't think  marking `maintains_input_order` to be `true` makes logical sense and marking  `maintains_input_order`  `false` means the optimizer will have maximum flexibility (as in it doesn't have to retain ordering). 

However, perhaps you are  imagining a case like

```sql
-- user is trying to  specify that `new_table` should be ordered by `x`
INSERT INTO new_table SELECT * from old_table ORDER BY x
```
Trying to use the order by of the input as a hint might cause problems with tables that already have data such as 


```sql
INSERT INTO new_table SELECT * from old_table ORDER BY x;
-- A second insert of the same data means new_table won't be 
-- sorted by x unless extra care is taken somehow within new_table
INSERT INTO new_table SELECT * from old_table ORDER BY x;
```

I was thinking that eventually the table provider for `new_table` would somehow set `required_input_order()` to be `x` in this case. Perhaps something like 

```sql
CREATE TABLE new_table ORDER BY x;
-- new_table would specify "required_input_order = x" 
-- properly handle merging with existing data 
INSERT INTO new_table SELECT * from old_table;
```

Reply:
I agree that ordering information is specific to the target table, and ordering of an individual write does not imply an ordering of the target table on its own.

However, I think writers should preserve the ordering they receive and write in that order, if there is any. I think a user writing
```sql
INSERT INTO new_table SELECT * from old_table ORDER BY x
```
and seeing a different order on the table afterwards would be bad UX.

Reply:
I did some investigation, and I am not sure the code needs to be changed. The code in this PR gets the desired behavior of preserving the order of the input query, as can be seen in the explain plan tests (I annotated with `****`) 

This query has `ORDER BY c1`
```sql
EXPLAIN
INSERT INTO table_without_values SELECT
SUM(c4) OVER(PARTITION BY c1 ORDER BY c9 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING),
COUNT(*) OVER(PARTITION BY c1 ORDER BY c9 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING)
FROM aggregate_test_100
ORDER by c1
```

The plan shows the ordering is maintained:
```
logical_plan
Dml: op=[Insert] table=[table_without_values]
--Projection: SUM(aggregate_test_100.c4) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING AS field1, COUNT(UInt8(1)) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING AS field2
----Sort: aggregate_test_100.c1 ASC NULLS LAST
------Projection: SUM(aggregate_test_100.c4) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING, COUNT(UInt8(1)) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING, aggregate_test_100.c1
--------WindowAggr: windowExpr=[[SUM(aggregate_test_100.c4) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING, COUNT(UInt8(1)) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING]]
----------TableScan: aggregate_test_100 projection=[c1, c4, c9]
physical_plan
InsertExec: sink=MemoryTable (partitions=1)
--ProjectionExec: expr=[SUM(aggregate_test_100.c4) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING@0 as field1, COUNT(UInt8(1)) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING@1 as field2]
----SortPreservingMergeExec: [c1@2 ASC NULLS LAST]
   ^******** You can see here the data remain sorted as it is fed into the InsertExec
------ProjectionExec: expr=[SUM(aggregate_test_100.c4) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING@3 as SUM(aggregate_test_100.c4), COUNT(UInt8(1)) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c9 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING@4 as COUNT(UInt8(1)), c1@0 as c1]
--------BoundedWindowAggExec: wdw=[SUM(aggregate_test_100.c4): Ok(Field { name: "SUM(aggregate_test_100.c4)", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }, COUNT(UInt8(1)): Ok(Field { name: "COUNT(UInt8(1))", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)) }], mode=[Sorted]
----------SortExec: expr=[c1@0 ASC NULLS LAST,c9@2 ASC NULLS LAST]
------------CoalesceBatchesExec: target_batch_size=8192
--------------RepartitionExec: partitioning=Hash([Column { name: "c1", index: 0 }], 8), input_partitions=8
----------------RepartitionExec: partitioning=RoundRobinBatch(8), input_partitions=1
------------------CsvExec: file_groups={1 gr
```

Reply:
That's good! Given that it exhibits the expected behavior, what do you think about having `maintains` return `true`? IMO it would be good for two reasons: (1) self-documentation and consistency, (2) future-proofing -- a future optimizer will likely not break this behavior given that return value.

Reply:
>  what do you think about having maintains return true?

The reason I am pushing back on changing the `maintains_input_order` is  that I don't think `InsertExec` operator preserves the input's order. Specifically it doesn't satisfy  the meaning of this setting 

https://github.com/apache/arrow-datafusion/blob/7e5c92db5f0dcc8261a27846165236c4d44460de/datafusion/core/src/physical_plan/mod.rs#L150-L165 

I guess you could argue that since the output is a single row (the count) it is technically  ordered in whatever way you want, but I still feel is it incorrect and unecessary

Maybe we can find some test / example where having `maintains_input_order` set to true makes a difference? If so that I will gladly change it

Reply:
This is more about future-proofing. We may need to add `InsertExec` as a special case in any future rules we write that may change sorts/orderings. Assume that `InsertExec` is at the top, and there is a `SortExec` under it. A rule that doesn't know the special "status" of `InsertExec` could say "The ordering imposed by this `SortExec` is lost after the `InsertExec`, and the `InsertExec` itself doesn't require this ordering either, so I may just remove the `SortExec` -- which would be wrong.

Reply:
Let's analyze a concrete case, say your example query. In that query, the ordering imposed by the `SortExec` in the middle of the plan is required by the `BoundedWindowAggExec` above it, which preserves that ordering. So far so good. However, isn't the optimizer then free to use `CoalescePartitionsExec` instead of `SortPreservingMergeExec` up top (if it thinks using the former is faster)? Where does the need to still preserve the ordering come from?

If it did that, we'd have a problem.

Reply:
> Would adding something like

Yes, thank you @ozankabak -- that makes sense and I have made that change and added a test in https://github.com/apache/arrow-datafusion/pull/6354/commits/7c2f18989ae70a274bad0521810056f754788327